File: app/src/main/java/com/tcc/sicv/data/model/FlowState.java
package com.tcc.sicv.data.model;

public class FlowState<T> {
    private T data = null;
    private Throwable throwable = null;
    private Status status = Status.NEUTRAL;

    public T getData() {
        return data;
    }

    public Throwable getThrowable() {
        return throwable;
    }

    public Status getStatus() {
        return status;
    }

    public FlowState(T data, Throwable throwable, Status status){
        this.data = data;
        this.throwable = throwable;
        this.status = status;
    }

    public <T> FlowState(){}

    public void loading(){
        data = null;
        throwable = null;
        status = Status.LOADING;
    }

    public Boolean isLoading(){ return status == Status.LOADING; }

    public Boolean hasData(){ return data != null; }
}

File: app/src/main/java/com/tcc/sicv/data/model/FlowState.kt
package com.tcc.sicv.data.model

import com.tcc.sicv.data.model.Status.*

data class FlowState<T>(
        val data: T?,
        val throwable: Throwable?,
        val status: Status
) {
    val isLoading: Boolean?
        get() = status == Status.LOADING

    companion object {
        inline fun <reified T> loading() = FlowState<T>(null, null, LOADING)
        inline fun <reified T> success(data: T) = FlowState(data, null, SUCCESS)
        inline fun <reified T> failure(e: Throwable?) = FlowState<T>(null, e, ERROR)
    }
}
